sql/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectExecutePlanHandler.scala

```diff
@@ -35,19 +35,38 @@ class SparkConnectExecutePlanHandler(responseObserver: StreamObserver[proto.Exec
       .getOrCreateIsolatedSession(v.getUserContext.getUserId, v.getSessionId, previousSessionId)
     val executeKey = ExecuteKey(v, sessionHolder)
 
-    SparkConnectService.executionManager.getExecuteHolder(executeKey) match {
-      case None =>
+    val executeHolder = SparkConnectService.executionManager.getExecuteHolder(executeKey)
+    val opStatus = sessionHolder.getOperationStatus(executeKey.operationId)
+    val abandoned = SparkConnectService.executionManager.getAbandonedTombstone(executeKey)
+
+    (executeHolder, opStatus, abandoned) match {
+      case (None, None, _) =>
         // Create a new execute holder and attach to it.
         SparkConnectService.executionManager
           .createExecuteHolderAndAttach(executeKey, v, sessionHolder, responseObserver)
-      case Some(executeHolder) if executeHolder.request.getPlan.equals(v.getPlan) =>
+      case (Some(executeHolder), _, _) if executeHolder.request.getPlan.equals(v.getPlan) =>
         // If the execute holder already exists with the same plan, reattach to it.
         SparkConnectService.executionManager
           .reattachExecuteHolder(executeHolder, responseObserver, None)
-      case Some(_) =>
+      case (Some(_), _, _) =>
+        logInfo(log"ExecuteHolder ${MDC(LogKeys.EXECUTE_KEY, executeKey)} exists.")
+        throw new SparkSQLException(
+          errorClass = "INVALID_HANDLE.OPERATION_ALREADY_EXISTS",
+          messageParameters = Map("handle" -> executeKey.operationId))
+      case (None, Some(_), None) =>
```

Comment:
Should we have a different error class for this?

Reply:
At the moment, we don't have any special handling for this error on the client.

I'm indifferent - do you see value in a separate code?